### PR TITLE
51660 Adds short-circuit filter to skip query for `WP_List_Table::months_dropdown()`

### PR DESCRIPTION
**Pull request by hellofromtonya**
Trac ticket: https://core.trac.wordpress.org/ticket/51660

Updates PR #767 to modify the added filter to be a short-circuit for the query.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-admin/includes/class-wp-list-table.php [hunk 1]
```diff
@@ -602,18 +602,31 @@ protected function months_dropdown( $post_type ) {
...
-		);
+		/**
+		 * Filters whether to short-circuit performing the months dropdown query.
+		 *
+		 * @since 5.7.0
+		 *
+		 * @param bool   $skip_query Whether to skip doing the query. Default false.
+		 * @param string $post_type  The post type.
+		 */
+		$skip_query = apply_filters( 'pre_months_dropdown_query', false, $post_type );
```
Comment (TimothyBJacobs) [inline src/wp-admin/includes/class-wp-list-table.php:613]: Hm, so I was expecting this to return the list of months if they want to short-circuit the filter. Like we do in `get_option` or `get_blogs_of_user` for instance.

Comment (hellofromtonya) [reply src/wp-admin/includes/class-wp-list-table.php:613]: There's an existing filter just below for filtering the results:

```
$months = apply_filters( 'months_dropdown_results', $months, $post_type );
```

The new filter short-circuits to skip the query and the existing filter allows you to filter the months.

What do you think @TimothyBJacobs ?

Comment (hellofromtonya) [reply src/wp-admin/includes/class-wp-list-table.php:613]: A preemptive filter already exists to disable the dropdown completely. It's a guard clause at the opening of the method:

https://github.com/WordPress/wordpress-develop/blob/ea4841d30e40e3f049ca3403a480bb6e10420c89/src/wp-admin/includes/class-wp-list-table.php#L583-L596

Comment (hellofromtonya) [reply src/wp-admin/includes/class-wp-list-table.php:613]: Re-reading the ticket's proposed enhancement, it seeks to give a way to skip the months query and allow a filtered months instead similar to https://core.trac.wordpress.org/browser/tags/5.5.1/src/wp-includes/media.php#L4106

Comment (TimothyBJacobs) [reply src/wp-admin/includes/class-wp-list-table.php:613]: Agreed that is what they are looking for. If we are following that pattern, it'd use the results returned from the pre-filter.

Comment (hellofromtonya) [reply src/wp-admin/includes/class-wp-list-table.php:613]: So you're proposing the new filter shifts the building of the dropdown's `label` and `select` to the 3rd party. Right?

Comment (hellofromtonya) [reply src/wp-admin/includes/class-wp-list-table.php:613]: Sorry for the confusion @TimothyBJacobs. 

- Are you suggesting that it bails out and returns the dropdown HTML?
If yes, then we shifting building the dropdown's `label` and `select` to the 3d party callback. Is this what we want here?

- Or that it's used to filter the months and uses the returned months (if any is returned)? 
If yes, a filter already exists for filtering the months. 

https://github.com/WordPress/wordpress-develop/blob/b6a04b3b28307bdedd34395b56cb140536257775/src/wp-admin/includes/class-wp-list-table.php#L618-L626

Comment (TimothyBJacobs) [reply src/wp-admin/includes/class-wp-list-table.php:613]: The second one, but the results filter shouldn't apply in that case. Having a double filter there is pretty typical.

Comment (hellofromtonya) [reply src/wp-admin/includes/class-wp-list-table.php]: Makes sense. Captured in commit d500f96be781c33858dd7a35c216bef834875db0.

File: src/wp-admin/includes/class-wp-list-table.php [hunk 2]
```diff
@@ -595,35 +595,47 @@ protected function months_dropdown( $post_type ) {
...
-		$months = apply_filters( 'months_dropdown_results', $months, $post_type );
+		$months = apply_filters( 'pre_months_dropdown_query', false, $post_type );
+
+		if ( ! is_array( $months ) ) {
+			$extra_checks = "AND post_status != 'auto-draft'";
+			if ( ! isset( $_GET['post_status'] ) || 'trash' !== $_GET['post_status'] ) {
+				$extra_checks .= " AND post_status != 'trash'";
+			} elseif ( isset( $_GET['post_status'] ) ) {
+				$extra_checks = $wpdb->prepare( ' AND post_status = %s', $_GET['post_status'] );
+			}
```
Comment (hellofromtonya) [inline src/wp-admin/includes/class-wp-list-table.php:614]: Moved the extra checks into the conditional branch as these only need to run if the query will run.

File: src/wp-admin/includes/class-wp-list-table.php [hunk 3]
```diff
@@ -595,35 +595,47 @@ protected function months_dropdown( $post_type ) {
...
+
+			/**
+			 * Filters the 'Months' drop-down results.
+			 *
+			 * @since 3.7.0
+			 *
+			 * @param object[] $months    Array of the months drop-down query results.
+			 * @param string   $post_type The post type.
+			 */
+			$months = apply_filters( 'months_dropdown_results', $months, $post_type );
```
Comment (hellofromtonya) [inline src/wp-admin/includes/class-wp-list-table.php:637]: Moved this filter into the conditional branch as it only needs to run if the query happened.

File: src/wp-admin/includes/class-wp-list-table.php (hunk 1 above)
Comment (TimothyBJacobs) [reply src/wp-admin/includes/class-wp-list-table.php]: LGTM!

Comment (hellofromtonya) [reply src/wp-admin/includes/class-wp-list-table.php]: @TimothyBJacobs Thanks for your guidance (and patience)!

**Comment by hellofromtonya** [comment]
Closed with changeset https://core.trac.wordpress.org/changeset/50163

**Pull request closed without merging**